### DISCUSSION TITLE
Why so many change types for mutable vec?

### DISCUSSION THREAD

**Discussion by andrewbaxter** [Pauan/rust-signals#70] [Q&A] [answered]
I think all vec changes can be represented as a single `splice` tuple with an offset, remove count, and add list.

I was wondering why this has so many different vec change event types (push, pop, insert, etc).

**Comment by Pauan** [comment, marked as answer]
That would be less efficient, because `Vec` requires a heap allocation. The only time a heap allocation is necessary with `VecDiff` is with the `Replace` variant, which completely replaces all the old values with new values.

`VecDiff` is an internal implementation detail, so the primary goal is correctness, functionality, and performance. Simplicity is a nice bonus, but it is less important than those other goals.

**Reply by andrewbaxter** [reply]
Ah I see, and it only allows single-element changes (no extend/range delete).  Thanks for answering!

**Reply by Pauan** [reply]
Yes, multiple changes are represented as multiple `VecDiff`.

Perhaps in the future I can add a new `Splice` variant that would use the [`Vec::splice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.splice) method.

In practice I've never used `Vec::splice`, it doesn't seem to be very common.
